common/flatpak-instance.c

```diff
@@ -424,6 +472,8 @@ flatpak_instance_new (const char *dir)
                                                     FLATPAK_METADATA_GROUP_INSTANCE, FLATPAK_METADATA_KEY_RUNTIME_COMMIT, NULL);
     }
 
+  priv->run_environ = get_run_environ (priv->private_dir);
```

Comment:
Splitting this comment by @smcv from [this thread](https://github.com/flatpak/flatpak/pull/5510/files#r1321787694) to its own separate thread:
> ...
>> As a side-note, I prefer to abort with an error in case of bad formatting than silently ignore environment variables and result in slightly broken behavior
>
>That's fair. I'd be OK with having this fail recoverably, but that would require passing a GError up to a layer where failing is reasonable to do. However, it is not OK for this to abort() or g_error(), because FlatpakInstance is public API in libflatpak (via flatpak_instance_get_all()). For example, it would not be OK for flatpak ps to crash just because there's an app running that was started before upgrading Flatpak.
>
>It's a bit awkward that this is a private method on a public object. We're constrained by FlatpakInstance not being GInitable (and retroactively making it GInitable would arguably be an incompatible change), which means its constructor can't fail; but if we load the environment block in the constructor, then that means failure to load the environment can't be signalled properly.
>
>Can we, instead, make this
>``` c
>const char * const *flatpak_instance_get_run_environ (FlatpakInstance *self, GError **error);
>```
>with the environment block loaded on-demand and cached in self after the first successful load; or even
>``` c
>GStrv flatpak_instance_get_run_environ (FlatpakInstance *self, GError **error);
>```
>with it loaded every time (no difference, in practice)? That way, if loading the environment block fails, the portal can do something like (pseudocode)
>``` c
>if (flatpak_instance_get_run_environ () succeeds)
>  use its result
>else if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
>  g_info(), use g_get_environ ()
>else
>  raise the error, return failure
> ```

Reply:
>It's a bit awkward that this is a private method on a public object. ...

While I don't disagree that it's a bit awkward, `FlatpakInstance` does seem to be the right object to hold the environment information.

> We're constrained by FlatpakInstance not being GInitable (and retroactively making it GInitable would arguably be an incompatible change), which means its constructor can't fail; but if we load the environment block in the constructor, then that means failure to load the environment can't be signalled properly.

I would have to guess that the design choice here was something along the lines of:
1. `flatpak_instance_new` would always seemingly succeed, even if `get_pid`, `get_child_pid` and `get_instance_info` (or others) would actually run into nonrecoverable errors.
2. The user would be informed about the relevant nonrecoverable errors later, by getting `NULL` or `0` from `flatpak_instance_get_pid`, `flatpak_instance_get_child_pid` and the rest of the `flatpak_instance_get_*` APIs.

In that sense, it seems `FlatpakInstance` abstracts away the implementation details of how the information it holds is obtained (disk? memory? DBus?) and when (during construction? when calling the individual `flatpak_instance_get_*` APIs? periodically every 5 seconds?).

As for the actual implementation, it seems that it was aimed at providing an object which represents the instance in a "frozen" state in time; repeated calls to `flatpak_instance_get_*` would always return the exact same result. It does sound sensible.

In fact, I feel that it makes sense to inherit the "frozen" trait of this implementation in the context of `flatpak_instance_get_run_environ`, so I personally don't see this as an issue.

Reply:
Looking at the latest revision of this PR, I see your reasoning, but I'm going to overrule you on this.

The simplest way to get the behaviour change you're looking for here will be to stop caching the run environment in the `FlatpakInstance`, and move the call to `get_run_environ()` (or the whole implementation of `get_run_environ()`, which doesn't necessarily need to be a separate function) into the implementation of `flatpak_instance_get_run_environ()`, with this signature:

```
GStrv flatpak_instance_get_run_environ (FlatpakInstance *self, GError **error);
```

returning a non-null array on success, or `NULL` (setting `error`) on error. That way, the error message bubbles up to a module that is sufficiently high-level to make a rational decision about "how bad is this?" and report it accordingly.

> As for the actual implementation, it seems that it was aimed at providing an object which represents the instance in a "frozen" state in time; repeated calls to flatpak_instance_get_* would always return the exact same result.

Yes ish, but the difference between "frozen snapshot" and "cache" is merely academic when the facts we're retrieving are things that by design can't change anyway, which, as far as I can see, is true for all of them.

In particular, the environment with which `flatpak run` was invoked is, by definition, a constant: it's a fact about the past, not a fact about the present. Whether we cache it in memory or re-load it every time, there is no situation in which it could validly change (other than the instance disappearing entirely, at which point obviously the stored environment disappears too).

In this case, it is not going to be used repeatedly within one process: it's a private interface (so we control all callers), and our only caller uses it once before throwing away the FlatpakInstance (so a cache would not give us any practical benefit). So we lose nothing by loading it lazily, on-demand.

Conversely, if we load it eagerly, we'll be wasting a bit of memory and I/O in every user of the public FlatpakInstance API to retrieve information that there's no public API to retrieve; and to let a caller of the private interface get at the error information, we'd have to cache a GError in the object, too, increasing the wasted memory. So, no, I don't think loading it eagerly is the right answer.

Reply:
Done per your suggestion.